## Coding our first neurons
#### A single neuron 

In [1]:
inputs = [1,2,3]
weights = [0.2,0.8,-0.5] 
bias = 2 

In [4]:
output = inputs[0]*weights[0] + inputs[1]*weights[1] + inputs[2]*weights[2] + bias

In [5]:
print(output)

2.3


### Visualization for the above code 
https://nnfs.io/bkr

In [6]:
## Let's change to 4 inputs 
inputs = [1.0,2.0,3.0,2.5] 
weights = [0.2,0.8,-0.5,1.0] 
bias = 2 

In [10]:
output = (inputs[0]*weights[0] + inputs[1]*weights[1] + inputs[2]*weights[2] + inputs[3]*weights[3] + bias) 

In [11]:
print(output)

4.8


![alt text](neurons.png "Title")

### Layers of Neurons 
![alt text](Neurons2.png "Title")

In [15]:
inputs = [1,2,3,2.5] 

weights1 = [0.2,0.8,-0.5,1]
weights2 = [0.5,-0.91,0.26,-0.5]
weights3 = [-0.26,-0.27,0.17,0.87] 
bias1 = 2 
bias2 = 3 
bias3 = 0.5 


In [16]:
output = [
    #Neuron 1: 
    inputs[0]*weights1[0] + 
    inputs[1]*weights1[1] + 
    inputs[2]*weights1[2] +
    inputs[3]*weights1[3]+ bias1,

    #Neuron 2 : 
    inputs[0]*weights2[0] + 
    inputs[1]*weights2[1] + 
    inputs[2]*weights2[2] +
    inputs[3]*weights2[3] + bias2,

    #Neuron 3 : 
    inputs[0]*weights3[0] +
    inputs[1]*weights3[1] + 
    inputs[2]*weights3[2] + 
    inputs[3]*weights3[3] + bias3 


]

In [17]:
print(output)

[4.8, 1.21, 2.385]


1. **Neuron 1 will provide output of 4.8**
2. **Neuron 2 will provide output of 1.21**
3. **Neuron 3 will provide output of 2.385** 
* This can serve as input for the neurons for the next layers 

Visuals for above code : https://nnfs.io/mxo

## This is the example of fully connected neural network 

In [4]:
inputs = [1,2,3,2.5] 
weights = [[0.2,0.8,-0.5,1],
           [0.5,-0.91,0.26,-0.5],
           [-0.26,-0.27,0.17,0.87]] 
biases = [2,3,0.5] 

# Output of current layers 
layers_output = []
# For each neurons
# zip pairs list of weights and its corresponding biases
for neuron_weight, neuron_bias in zip(weights,biases):
    #Zeroed output for given neuron 
    neuron_output = 0 
    # For each input and weight to the neuron
    #zip pairs each input with its corresponding weights
    for n_input,weight in zip(inputs,neuron_weight):
        # Multiply this input by associated weight
        # and add to the neuron's output variable 
        neuron_output+= n_input*weight
    
    
    #Add bias 
    neuron_output += neuron_bias

    
    # Put neuron's result to the layer's output list 
    layers_output.append(neuron_output)

print(layers_output)


[4.8, 1.21, 2.385]


In [6]:
layers_output=[]
#for each neuron 
for neuron_weight, neuron_bias in zip(weights,biases):
    print(neuron_weight)
    print(neuron_bias)

[0.2, 0.8, -0.5, 1]
2
[0.5, -0.91, 0.26, -0.5]
3
[-0.26, -0.27, 0.17, 0.87]
0.5


In [1]:
list = [1,2,3,4]

In [2]:
list[1]

2

In [3]:
lol = [[1,2],
       [3,4],
       [5,6]]

In [10]:
lol[1][0]

3

###  multi dimensional lists and accesing elements within them 

In [14]:
lolol = [[[1,5,6,2],
        [3,2,1,3]],
        [[5,2,1,2],
         [6,4,8,4]],
         [[2,8,5,3],
          [1,1,9,4]]]

In [16]:
lolol[0][0][3]

2

### Vector multiplication or dot product 

In [17]:
a = [1,2,3] # input
b = [2,3,4] # weights

In [18]:
# To obtain dot product 
dot_product = a[0]*b[0] + a[1]*b[1] + a[2]*b[2]

In [19]:
print(dot_product)

20


#### Let's do dot product using numpy 


In [25]:
import numpy as np 
inputs = [1.0, 2.0, 3.0, 2.5]
weights = [0.2, 0.8, -0.5, 1.0]
bias = 2.0

In [26]:
output = np.dot(weights,inputs) + bias 

In [27]:
print(output)

4.8
